langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiEmbeddingModel.java

```diff
@@ -82,9 +111,11 @@ public Response<List<Embedding>> embedAll(List<TextSegment> segments) {
             List<Embedding> embeddings = new ArrayList<>();
             int inputTokenCount = 0;
 
-            for (int i = 0; i < segments.size(); i += BATCH_SIZE) {
+            List<Integer> tokensCounts = getTokensCounts(segments);
```

Comment:
This step seems to add around 30% to the overall latency :( Would be nice if google could publish their tokenizer algo so that we could count tokens without the need to do network calls.

For now, I am wondering if we could approximate the calculation of tokens somehow? For the general use case. User could still configure it to use the proper estimation, but we could use a faster estimate instead of a slow but precise calculation by default.
2 options that come to mind:
- count number of words (by roughly splitting by space). it seems that 1 english word is rougly 1.2-1.3 tokens, so we can play safe and use 1.
-  count number of chars. it seems that 1000 chars is rougly 200-240 tokens, so we could play safe and use 180 or so.
WDYT?

Reply:
Before sending this PR, I actually went through different prototypes.

Since I couldn't find a built-in function that would calculate token counts (without calling the server), I went with the approach of estimating the token counts, but even with a conservative approach, on some dummy text, I hit the limit. So I started to loop till I hit a token limit exception, with some splitting/dichotomy approach: let's try to see if this can handle everything, ah no, then split in two, and do two calls, ah one call fails, let's split again.
But in the end, all these tricks seemed a bit over-engineered, and a bit ugly to my taste.

So instead, I went the route of calling the API that counts token. It's not ideal, and adds overhead, but the code ended being a bit more elegant, and hopefully easier to maintain in the long run. Also, indeed making those calls add 30%, but although I haven't measured the overhead with my other approach (dichotomy + error handling), I had some overhead as well as I asked, say, to embed 250 segments, boom it fails, let's try again with 125 and 125, boom one fail, let's try 64 and 64, etc.

A strange oddity though is that counting and summing up individual text segments tokens doesn't necessarily gives the same number as when I send the all the segments together, which is a bit unintuitive, and I haven't found why, but kept the choice to specify limits (max number of segments, max number of tokens)

Now, that said, I have some hope to convince the people working on the models to share their embedding algorithm somehow, and if they ever do, we'll be able to replace the network API call very easily with the library code that would calculate the number of tokens.

So I think I'll stick to my current approach in this PR, unless you strongly disagree.

Reply:
Ok, makes sense! thanks for explanation